Adding FWH plugin

WillTrojak opened the pull request:
This is a spatial domain FWH solver, which greatly reduces the overhead and doesn't require special handling of restarts compared to frequency domain FWH.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,822 @@
+from collections import defaultdict
+import re
+
+from mpi4py import MPI
```
Don't import MPI will break PyFR.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,822 @@
+            wdata = wa_arr
+            obploc = np.array([opts.tolist()] * nfreq)
+
+            for op in [*obploc.T]:
```
This will be slow

WillTrojak replied:
Ok, I'll have a look at refactoring this.

mhawwary commented:
Hi @WillTrojak , thanks for revisiting this work. I think it is better to work on the other repo that I added you too. The branch name is "features/fwhplugin". This is the recent one so will save you some time as it already incorporates some of Freddie's suggestions and comments. It also uses the latest region class that Freddie added last year, does not import MPI, etc.

WillTrojak commented:
@mhawwary I have moved the region stuff over now and changed the MPI. I think it's best to keep working on this version as I've done significant refactoring already. Let me know if there are any other modifications though that you think would be appropriate.

WillTrojak commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,785 @@
+
+        e, self._int_eidxs = self._int_faces(mesh, rallocs, surfeset, sfaces,
+                                             periodic, self.bcs)
+        dd_list_merge(self._eidxs, e)
```
I think this can be simplified, _eidxs doesn't get used anymore in the generation of the various faces and is now only a union of _int_eidxs, _mpi_eidx, _bnd_eidxs

mhawwary replied:
> @mhawwary I have moved the region stuff over now and changed the MPI. I think it's best to keep working on this version as I've done significant refactoring already. Let me know if there are any other modifications though that you think would be appropriate.

That seems good. It looks like you started the refactoring on a version that already have most of what is in that private repo. I did not notice that in the beginning. I will let you know if I have any comments but looks good so far. Thanks for working on this.

mhawwary commented:
> @mhawwary I have moved the region stuff over now and changed the MPI. I think it's best to keep working on this version as I've done significant refactoring already. Let me know if there are any other modifications though that you think would be appropriate.

That seems good. It looks like you started the refactoring on a version that already have most of what is in that private repo. I did not notice that in the beginning. I will let you know if I have any comments but looks good so far. Thanks for working on this.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+
+        if self.incomp:
+            self.qinf['rho'] = self.uinf['p']/self.consts['ac-zeta']
+            self.qinf['c'] = np.sqrt(self.consts['ac-zeta'])
```
Usually **0.5 is preferred over sqrt.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        else:
+            gamma = self.consts['gamma']
+            self.qinf['c'] = np.sqrt(gamma * self.qinf['p'] / self.qinf['rho'])
+        self.qinf['M'] = [self.qinf[k]/self.qinf['c'] for k in 'uvw'[:ndims]]
```
Newline.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        self.Minf = np.linalg.norm(self.qinf['M'], ord=2)
+
+        ele_map = intg.system.ele_map
+        self.emap = {k: i for i, k in enumerate(ele_map.keys())}
```
No need for .keys().

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+
+                m0 = eles.basis.m0[facefpts]
+                a = eles.basis.fpts_wts[facefpts][:, None] / rcpdjacs[fslice]
+                a = np.reshape(a, (-1,))
```
Prefer `a.reshape` (the module level functions are used when we are not sure that `a` is a NumPy array).

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        s = np.reshape(spts, (-1, self.ndims))
+        o = self.obsv_pts
+
+        r_o = o[None, :, :] - s[:, None, :]
```
Can remove final `:`.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+            r_tot = r_inf + p * ci**2
+
+            mom = r_tot * (u + self.uinf[:, None])
+            drift = - r_inf * self.uinf[:, None]
```
Remove space before unary -.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        o_vals = np.zeros(self.nobvs)
+        ci = 1 / self.qinf['c']
+
+        for (etype, fidx), p in self.proj.items():
```
Ensure that all soln-like properties are queried outside of the loop to prevent MPI deadlocks.

FreddieWitherden replied:
You can just do `soln = intg.soln` and `soln_t = intg.dt_soln` at the top of the function.  This ensures there will be no divergence.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        if intg.tcurr < self.tstart:
+            return
+
+        comm, rank, root = get_comm_rank_root()
```
Is this needed?

FreddieWitherden commented on pyfr/solvers/euler/elements.py:
```diff
@@ -47,6 +47,26 @@ def con_to_pri(cons, cfg):
+                               np.einsum('ijk,ijk->jk', rhouvw, diff_uvw) )
+        diff_p *= (gamma - 1)
+
+        return [diff_rho] + diff_uvw + [diff_p]
```
I think we can now do `[diff_rho, *diff_uvw, diff_p]`

FreddieWitherden commented on pyfr/integrators/dual/phys/base.py:
```diff
@@ -49,6 +49,22 @@ def grad_soln(self):
+            copy = self.soln
+            idx = self.pseudointegrator._idxcurr
+            system.rhs(self.tcurr, idx, idx)
+            self._curr_dt_soln = system.ele_scal_upts(self._idx)
```
Do we want self._idx here?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        self.t_window = self.cfg.getfloat(cfgsect, 'tend', intg.tend)
+        self.obsv_pts = np.array(self.cfg.getliteral(self.cfgsect,
+                                                     'observer-pts'))
+        self.nobvs = self.obsv_pts.shape[0]
```
len

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        o = self.obsv_pts
+
+        r_o = o[None, :, :] - s[:, None, :]
+        d = np.linalg.norm(r_o, ord=2, axis=-1)
```
No need for `ord=2` as it is the default.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+
+        m_r = self.qinf['M'] * x
+        m_r_dot = np.sum(m_r, axis=-1)
+        m_inf = np.linalg.norm(self.qinf['M'], ord=2, axis=0)
```
If we are just going to square it I think dot may be what you want?

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -140,6 +140,8 @@ def __init__(self, intg, *args, **kwargs):
                 rgn = BoundaryRegion(m[1], nlayers=int(m[2] or 1))
 
             eset = rgn.interior_eles(intg.system.mesh, intg.rallocs)
+            self._surf_region_data = rgn.surface_faces(intg.system.mesh,
```
Add a method to handle this, so only people that want the faces get it?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+            I += ci * r_star_inv * np.sum(F_t[...,None] * R.transpose(2, 0, 1),
+                                          axis=0)
+            I += (np.sum(F[...,None] * R_star.transpose(2, 0, 1), axis=0) *
+                  r_star_inv * r_star_inv)
```
Use **2 for squaring.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+            p_t = pris_t[self._pidx, ...]
+
+            r_tot_t = p_t / (self.qinf['c']**2)
+            mom_t = (r_tot_t * (u + self.uinf[:, None]) + r_tot * u_t)
```
Usually no space around * (but keep it for /).

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        ci = 1 / self.qinf['c']
+
+        for (etype, fidx), p in self.proj.items():
+            eidx, m0, n, a, (R, r_star_inv, R_star) = p
```
Different capitalisation?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+    systems = ['ac-euler', 'ac-navier-stokes', 'euler', 'navier-stokes']
+    formulations = ['dual', 'std']
+
+    def __init__(self, intg, cfgsect, suffix=None, *args, **kwargs):
```
Handle invalid regions * and boundaries.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+
+
+class FWHPlugin(RegionMixin, BaseSolnPlugin):
+    name = 'fwhsolver'
```
Naming convention (vs. class and Python file name).

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+
+from pyfr.mpiutil import get_comm_rank_root
+from pyfr.nputil import npeval
+from pyfr.plugins.base import BaseSolnPlugin, init_csv, RegionMixin
```
Ordering.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+
+        self.tstart = self.cfg.getfloat(cfgsect, 'tstart', 0.0)
+        self.dt = self.cfg.getfloat(cfgsect, 'dt')
+        self.t_window = self.cfg.getfloat(cfgsect, 'tend', intg.tend)
```
I don't think this is ever used?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        self.elementscls = intg.system.elementscls
+
+        self.tstart = self.cfg.getfloat(cfgsect, 'tstart', 0.0)
+        self.dt = self.cfg.getfloat(cfgsect, 'dt')
```
Are we aiming to exactly hit this dt?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+        self.uinf = np.array([self.qinf[k] for k in 'uvw'[:ndims]])
+
+        if self.incomp:
+            self.qinf['rho'] = self.uinf['p']/self.consts['ac-zeta']
```
This seems a bit oddball?  What do true incompressible codes with FWH do here?  Probing ac-zeta may be a bit fragile as it may be locally adaptive/variable in the future.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+            rcpdjacs = eles.rcpdjac_at_np('fpts')
+
+            for fidx in set(x for i, x in v):
+                eidx = [i for i, x in v if x == fidx]
```
Can these lines be cleaned up (iterating over v in a for loop which is over v).

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,180 @@
+            I += (np.sum(F[...,None] * R_star.transpose(2, 0, 1), axis=0) *
+                  r_star_inv * r_star_inv)
+            o_vals += a @ I
+        return o_vals / (4 * np.pi)
```
Add newline.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+
+                m0 = eles.basis.m0[facefpts]
+                a = eles.basis.fpts_wts[facefpts][:, None] / rcpdjacs[fslice]
+                a = a.reshape((-1,))
```
`a.flat` or `a.ravel()` will do it.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+                a = eles.basis.fpts_wts[facefpts][:, None] / rcpdjacs[fslice]
+                a = a.reshape((-1,))
+                n_mag = np.linalg.norm(pnorm[fslice], axis=-1, keepdims=True)
+                n = np.reshape((pnorm[fslice] / n_mag).transpose(2, 0, 1),
```
Can just do .reshape

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+        inv_gamma = np.sqrt(1 - self.Minf**2)
+
+        gamma = 1 / inv_gamma
+        s_fact = np.sqrt(inv_gamma*inv_gamma + m_r_dot*m_r_dot)
```
**0.5

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+
+    def _distances(self, etype, eles, fslice):
+        spts = eles.ploc_at_np('fpts').transpose(0, 2, 1)[fslice]
+        s = spts.reshape((-1, self.ndims))
```
No need for ().

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+            o_vals += a @ I
+        return o_vals / (4*np.pi)
+
+    def write_obvs(self, intg, o_vals):
```
Probably easiest to inline this as it is only called once.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+            u_t = pris_t[self._vidx]
+            p_t = pris_t[self._pidx]
+
+            r_tot_t = p_t / (self.qinf['c']**2)
```
No need for ().

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,196 @@
+
+            # Thickness noise contribution
+            I = (1 - self.Minf)*q_t[:, None]*r_star_inv
+            I -= (R_star @ self.uinf)*q[:, None]*r_star_inv**2
```
Would it be legitimate to multiply R_star through by r_star_inv into a temporary?

FreddieWitherden commented on doc/src/user_guide.rst:
```diff
@@ -1590,6 +1590,64 @@ Example::
+
+    *float*
+
+2. ``tend`` --- time at which to stop sampling, default is ``tend`` of time
```
I do not think this is used.

FreddieWitherden replied:
Do we need this? No other plugins have it?

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -155,3 +155,16 @@ def _prepare_region_data_eset(self, intg, eset):
             self._ele_regions.append((doff, etype, darr))
             self._ele_region_data[f'{etype}_idxs'] = darr
+
+    def prepare_surface_data(self, intg):
```
Another possibility is that we instead save the region instance, or if the region is '*' don't set it.  This will make it easy to query faces later on should someone need them.  Plus callers can then use isinstance to see if the region is geometric or not.

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -53,6 +53,9 @@ def __init__(self, backend, rallocs, mesh, initsoln, cfg):
         self._curr_grad_soln = None
 
+        # Solution dt cache
+        self._curr_dt_soln = None
```
I think it make sense to factor these out into a function, as currently we have them in three places (base, and accept step for both standard and dual controllers).

FreddieWitherden commented on doc/src/user_guide.rst:
```diff
@@ -1590,6 +1590,64 @@ Example::
+
+    [soln-plugin-fwhsolver]
+    file = fwh.csv
+    region = box((1,-5), (10, 5))
```
Spacing.

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -47,11 +47,7 @@ def __init__(self, backend, rallocs, mesh, initsoln, cfg):
-
-        # Solution gradients cache
-        self._curr_grad_soln = None
+        self.invalidate_caches()
```
Add comment

FreddieWitherden commented on pyfr/integrators/base.py:
```diff
@@ -122,6 +118,11 @@ def call_plugin_dt(self, dt):
             if not tlist or t - tlist[-1] > self.dtmin:
                 tlist.append(t)
 
+    def invalidate_caches(self):
```
Prefix with an _ since other people should not be calling this.

FreddieWitherden commented on pyfr/integrators/dual/phys/base.py:
```diff
@@ -49,6 +49,22 @@ def grad_soln(self):
         return self._curr_grad_soln
 
+    @property
+    def dt_soln(self):
```
Can we try and factor this method out across dual time and std? (Same goes for its friends above.)

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,197 @@
+import numpy as np
```
Rename to fwh.py?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,197 @@
+        self.uinf = np.array([self.qinf[k] for k in 'uvw'[:ndims]])
+
+        if self.incomp:
+            self.qinf['rho'] = self.cfg.getexpr(cfgsect, 'rho')
```
Does this need evaluating?

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,197 @@
+
+        if self.incomp:
+            self.qinf['rho'] = self.cfg.getexpr(cfgsect, 'rho')
+            self.qinf['c'] = self.cfg.getexpr(cfgsect, 'c')
```
Is this parameter documented?

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -124,20 +124,21 @@ def __init__(self, intg, *args, **kwargs):
             # Geometric region
             if '(' in region:
-                rgn = ConstructiveRegion(region)
+                self._rgn = rgn = ConstructiveRegion(region)
```
Make this public and give it a different name? Like `region_obj`?

FreddieWitherden commented on pyfr/solvers/euler/elements.py:
```diff
@@ -47,6 +47,26 @@ def con_to_pri(cons, cfg):
+        gamma = cfg.getfloat('constants', 'gamma')
+        diff_p = diff_E - 0.5*(np.einsum('ijk,ijk->jk', uvw,  diff_rhouvw) +
+                               np.einsum('ijk,ijk->jk', rhouvw, diff_uvw) )
+        diff_p *= (gamma - 1)
```
No need for ().

FreddieWitherden commented on pyfr/solvers/euler/elements.py:
```diff
@@ -47,6 +47,26 @@ def con_to_pri(cons, cfg):
+        # Pressure gradient: ∂p = (γ - 1)·[∂E - 1/2*(u⃗·∂(ρu⃗) - ρu⃗·∂u⃗)]
+        gamma = cfg.getfloat('constants', 'gamma')
+        diff_p = diff_E - 0.5*(np.einsum('ijk,ijk->jk', uvw,  diff_rhouvw) +
+                               np.einsum('ijk,ijk->jk', rhouvw, diff_uvw) )
```
Move space after ).

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,197 @@
+        self.t_window = self.cfg.getfloat(cfgsect, 'tend', intg.tend)
+        self.obsv_pts = np.array(self.cfg.getliteral(self.cfgsect,
+                                                     'observer-pts'))
+        self.nobvs = len(self.obsv_pts)
```
Not really needed.

FreddieWitherden commented on pyfr/plugins/acoustics.py:
```diff
@@ -0,0 +1,197 @@
+            # Loading noise contribution
+            I += ci*r_star_inv*np.sum(F_t[..., None]*R.transpose(2, 0, 1),
+                                       axis=0)
+            I += (np.sum(F[..., None]*R_star.transpose(2, 0, 1), axis=0)*
```
einsum might be better here

FreddieWitherden commented:
Would you be able to split the plumbing part out, as I think the `dt_soln` stuff is ready to merge.

FreddieWitherden commented:
Are you okay to rebase? And is there anything outstanding functionality or testing wise before we can merge?

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,197 @@
+            else:
+                comm.Reduce(mpi.IN_PLACE, o_vals, op=mpi.SUM, root=root)
+
+                self.write_obvs(intg, o_vals)
```
Does this exist?

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,197 @@
+
+        # Accumulate FWH contribution from each surface part
+        for (etype, fidx), p in self.proj.items():
+            eidx, m0, n, a, (R, r_star_inv, R_star) = p
```
Not sure if proj is the best variable name.  Also not sure if we need the second level of packing.

FreddieWitherden replied:
It might be worth stuffing these into either a nametuple or a dataclass.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,197 @@
+
+        soln, soln_t = {}, {}
+        # Querry solutions all at once to prevent MPI deadlock
+        for (etype, fidx), p in self.proj.items():
```
This can be simplified a little (just save soln and dt_soln into variables and merge the rest with the loop below).

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,197 @@
+        inv_gamma = np.sqrt(1 - self.Minf**2)
+
+        gamma = 1 / inv_gamma
+        s_fact = (inv_gamma*inv_gamma + m_r_dot*m_r_dot)**0.5
```
hypot

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -5,8 +5,11 @@
 from pyfr.mpiutil import get_comm_rank_root
+from pyfr.quadrules import get_quadrule
 from pyfr.regions import BoundaryRegion, ConstructiveRegion
+from pyfr.shapes import _proj_pts
```
Don't import _.  Should rename or move.

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+            q = get_quadrule(itype, qrule, qdeg=qdeg, flags='s')
+            ppts, pwts = _proj_pts(proj, q.pts), q.wts
+            self._surf_quad[itype, proj] = (ppts, pwts)
+        return ppts, pwts
```
Add newline.

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+        mag_pnorm = np.einsum('...i,...i', pnorm, pnorm)
+
+        # Check that none of these magnitudes are zero
+        if np.any(np.sqrt(mag_pnorm) < 1e-10):
```
Do we need to check this? It should be caught elsewhere.

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+        plocop = eles.basis.sbasis.nodal_basis_at(pts)
+
+        # Apply the operator to the mesh elements and reshape
+        e_surf = eles.eles[:,eidx]
```
Spacing after ,

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+
+        smats, djacs = eles._smats_djacs_mpts
+        djacs = djacs[:,eidx]
+        smats = smats[...,[e + neles*i for i in range(ndims) for e in eidx]]
```
Seems like there is a cleaner way to do this with reshape.

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+    def _jacs(eles, eidx, pts):
+        neles, ndims = eles.neles, eles.ndims
+
+        smats, djacs = eles._smats_djacs_mpts
```
Avoid accessing private members (or refactor if we have to so they are no longer private).

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+        npts = smats.shape[1]
+        smat = smats.transpose(1, 3, 0, 2)
+
+        norms = np.vstack([[fn]*n for fn, n in zip([norm], [npts])])
```
Not sure what is going on here with zip, but it seems broken.

WillTrojak replied:
This is taken from the norm code elsewhere. It just makes an array of norm vectors

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
+            self._surf_region_data = self._rgn.surface_faces(intg.system.mesh,
+                                                             intg.rallocs)
+
+class ElementMixin:
```
Use this for fluid force to ensure that it is flexible enough.

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -178,6 +195,22 @@ def __init__(self, intg, *args, **kwargs):
+
+        # Parse the region
+        sidxs = surface_data(self.cfg, self.cfgsect, intg.system.mesh,
+                            intg.rallocs)
```
Indentation.

FreddieWitherden commented on pyfr/solvers/base/elements.py:
```diff
@@ -217,7 +217,7 @@ def newfn(self, name, side=None):
 
     @memoize
     def smat_at_np(self, name):
-        smats_mpts, _ = self._smats_djacs_mpts
```
What are these changes needed for?

FreddieWitherden commented on pyfr/regions.py:
```diff
@@ -71,7 +71,7 @@ def surface_faces(self, mesh, rallocs, exclbcs=[]):
         nsfaces = defaultdict(list)
         for etype, eidx, fidx in sfaces:
-            nsfaces[etype].append((eidx, fidx))
+            nsfaces[etype,fidx].append(eidx)
```
Spacing.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,192 @@
+                    print(intg.tcurr, *x, p, sep=',', file=self.outf)
+
+                # Flush to disk
+                self.outf.flush()
```
Missing newline.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,192 @@
+            self.t_last = intg.tcurr
+
+            o_vals = self.fwh_solve(intg)
+            comm.Barrier()
```
Why is this needed?

FreddieWitherden replied:
Reduce includes a barrier.

FreddieWitherden replied:
It will be an MPI library specific thing, but the timeout is certainly not part of the MPI spec.  Also, we use a similar reduce call in several other plugins without issue.

FreddieWitherden replied:
Just following up here.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,192 @@
+        o_vals = np.zeros(self.nobvs)
+        ci = 1 / self.qinf['c']
+
+        # Querry dt_soln to prevent MPI deadlock
```
Query

FreddieWitherden commented on doc/src/user_guide.rst:
```diff
@@ -1604,6 +1604,66 @@ Example::
+    region = box((1,-5), (10, 5))
+    header = true
+    tstart = 10
+    tend = 100
```
Are we okay to remove?

FreddieWitherden replied:
Just following up here.

FreddieWitherden replied:
Just following up here.

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -178,6 +195,22 @@ def __init__(self, intg, *args, **kwargs):
 
 
 class SurfaceMixin:
+    def __init__(self, intg, *args, **kwargs):
```
Is this something we always want to do?  I do not think the fluid force needs this functionality, for example, yet it is in the constructor.

WillTrojak replied:
I can move it to another method, it doesn't depend on the `ele_map` so it should be fine to call anytime.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,192 @@
+            m0 = eles.basis.ubasis.nodal_basis_at(qpts)
+            norn_mag = np.linalg.norm(pnorm, axis=-1, keepdims=True)
+            n = (pnorm / norn_mag).transpose(2, 0, 1).reshape(self.ndims, -1)
+            area = (qwts[:, None] / rcpdjac).flatten()
```
Why not reshape?

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,192 @@
+            area = (qwts[:, None] / rcpdjac).flatten()
+            dist = self._distances(ploc)
+
+            self.surf[etype, fidx] = (eidxs, m0, n, area, *dist)
```
Use a namedtuple for this thing.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,192 @@
+        for (etype, fidx), p in self.surf.items():
+            eidx, m0, n, a, r_tilde_v, r_star_inv, r_star_tilde_v = p
+
+            soln = intg.soln[self.emap[etype]][..., eidx]
```
Shouldn't eidx be plural?

FreddieWitherden commented on pyfr/plugins/base.py:
```diff
@@ -178,6 +195,20 @@ def __init__(self, intg, *args, **kwargs):
+                             intg.rallocs)
+
+        # Generate the appropriate metadata arrays
+        self._ele_surface, self._ele_surface_data = [], {}
```
Return these as a tuple?

WillTrojak replied:
Return what as a tuple? This init is almost identical to the region mixini init

FreddieWitherden replied:
Have surf region return rather than set members.  It is more flexible and intuitive (call a function, get stuff back, rather than it implicitly setting members you need to query).

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,200 @@
+
+from pyfr.mpiutil import get_comm_rank_root, mpi
+from pyfr.nputil import npeval
+from pyfr.plugins.base import BaseSolnPlugin, init_csv, SurfaceMixin
```
Sort imports.

WillTrojak replied:
Are these not sorted?

FreddieWitherden replied:
```
>>> sorted(['BaseSolnPlugin', 'init_csv', 'SurfaceMixin'])
['BaseSolnPlugin', 'SurfaceMixin', 'init_csv']
```

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,200 @@
+            s_t = param.m0 @ soln_t.transpose(1, 0, 2)
+
+            pris = self.elementscls.con_to_pri(s, self.cfg)
+
```
Remove newline?

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,200 @@
+            mom_n = np.sum(param.norm*mom, axis=0, keepdims=True)
+            f = mom_n*u + (p - self.qinf['p'])*param.norm
+            f_t = mom_t_n*u + mom_n*u_t + p_t*param.norm
+            acc += ci*param.r_star_inv*np.einsum('ki,ijk->ij', f_t,
```
I think these lines can be cleaned up a little bit in terms of formatting.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,200 @@
+    formulations = ['dual', 'std']
+    dimensions = [2, 3]
+
+    SurfParams = namedtuple('SurfParams', ['eidxs', 'm0', 'norm', 'area',
```
Usually we define outside the class.

FreddieWitherden commented:
Also, how difficult would it be to add in quadrupole moments?

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,208 @@
+        if rank == root:
+            header = ','.join(['t', 'x', 'y', 'z'][:self.ndims + 1] + ['mag'])
+            self.outf = init_csv(self.cfg, cfgsect, header)
+            self.outf.flush()
```
Why do we need to flush here?

WillTrojak replied:
I think this was from when I ran it with a `tstart` and wanted to ensure it was doing something.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,208 @@
+                                   'uvw'[:ndims]])
+        self.Minf = np.linalg.norm(self.qinf['M'])
+        if self.Minf >= 1:
+            raise ValueError("FWH farfield Mach number greater than 1.")
```
Single quotes, no full stop.  Can use := for the assignment if you like.

WillTrojak replied:
I don't think you can use `:=` for an attribute of a class.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,208 @@
+
+        return r_tilde_vec, r_star_inv, r_star_tilde_vec
+
+    def fwh_solve(self, intg):
```
Should be _.

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,208 @@
+            q_t = mom_t_n.T
+            acc = (1 - self.Minf)*q_t*param.r_star_inv
+            acc -= q*param.r_star_inv**2*(
+                param.r_star_tilde_vec @ self.uinf.reshape((-1))
```
reshape(-1)

FreddieWitherden commented on pyfr/plugins/fwh.py:
```diff
@@ -0,0 +1,208 @@
+        dt_soln = intg.dt_soln
+
+        # Accumulate FWH contribution from each surface part
+        for (etype, fidx), param in self.surf.items():
```
Might be easier to index by eidx rather than etype.

WillTrojak commented:
> Also, how difficult would it be to add in quadrupole moments?

I haven't seen a derivation for a quadrapole in moving media so we might have to derive it as most people just do monopoles and dipoles. Also it involves terms $d^2/dx_idx_j$ so it might be a bit pricey to calculate.
